src/kde/quick_index.cc

```diff
@@ -0,0 +1,813 @@
+
+#include "quick_index.hh"
+
+namespace rtt_kde {
+//------------------------------------------------------------------------------------------------//
+/*!
+ * \brief
+ * 
+ * quick_index constructor. This function builds up a global indexing table to
+ * quickly access data that is spatial located near each other. It breaks up
+ * the data into equally spaced bins in each dimension. For domain decomposed
+ * data it builds a one sided communication map to place local data that is
+ * need on other processors for ghost cells. The ghost cell extents is
+ * determined by the max_data_window spatial size such that any cell on the
+ * local domain will have access to all points that should fall into the
+ * spatial window centered on any given local point.
+ *
+ * \tparam dim integer specifying the data dimensionality 
+ * \param[in] data locations
+ * \param[in] domain_decomposed
+ */
+quick_index::quick_index(const size_t dim_, const std::vector<std::array<double, 3>> &locations_,
+                         const double max_window_size_, const size_t bins_per_dimension_,
+                         const bool domain_decomposed_)
+    : dim(dim_), domain_decomposed(domain_decomposed_), coarse_bin_resolution(bins_per_dimension_),
+      max_window_size(max_window_size_), locations(locations_), n_locations(locations_.size()) {
+
+  // Build local bounding box
+  bounding_box_min = {0, 0, 0};
+  bounding_box_max = {0, 0, 0};
+  // only set initial values for working dimensions
+  for (size_t d = 0; d < dim; d++) {
+    bounding_box_min[d] = 1e20;
+    bounding_box_max[d] = -1e20;
+  }
+
+  for (auto locItr = locations.begin(); locItr != locations.end(); locItr++) {
+    auto &loc = *locItr;
+    for (size_t d = 0; d < dim; d++) {
+      if (loc[d] < bounding_box_min[d])
+        bounding_box_min[d] = loc[d];
+      if (loc[d] > bounding_box_max[d])
+        bounding_box_max[d] = loc[d];
+    }
+  }
+
+  if (domain_decomposed) {
+    // Store the local bounding box and extend to maximum non-local data size
+    local_bounding_box_min = bounding_box_min;
+    local_bounding_box_max = bounding_box_max;
+    for (size_t d = 0; d < dim; d++) {
+      local_bounding_box_min[d] -= max_window_size * 0.5;
+      local_bounding_box_max[d] += max_window_size * 0.5;
+    }
+    // Global reduce to get the global min and max
+    rtt_c4::global_min(&bounding_box_min[0], 3);
+    rtt_c4::global_max(&bounding_box_max[0], 3);
+    for (size_t d = 0; d < dim; d++) {
+      local_bounding_box_min[d] = std::max(local_bounding_box_min[d], bounding_box_min[d]);
+      local_bounding_box_max[d] = std::min(local_bounding_box_max[d], bounding_box_max[d]);
+    }
+  }
+
+  // temp cast corse_bin_resolution to double for interpolation
+  const double crd = static_cast<double>(coarse_bin_resolution);
+
+  // build up the local hash table of into global bins
+  size_t locIndex = 0;
+  for (auto locItr = locations.begin(); locItr != locations.end(); locItr++) {
```

Comment:
This could also be a range based for loop since it doesn't look like an iteration index is needed?